src/AspNet.Security.OAuth.Alipay/AlipayAuthenticationHandler.cs

```diff
@@ -0,0 +1,242 @@
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Net.Http;
+using System.Security.Claims;
+using System.Security.Cryptography;
+using System.Text;
+using System.Text.Encodings.Web;
+using System.Text.Json;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace AspNet.Security.OAuth.Alipay
+{
+    /// <summary>
+    /// Defines a handler for authentication using Alipay.
+    /// </summary>
+    public class AlipayAuthenticationHandler : OAuthHandler<AlipayAuthenticationOptions>
+    {
+        public AlipayAuthenticationHandler(
+            [NotNull] IOptionsMonitor<AlipayAuthenticationOptions> options,
+            [NotNull] ILoggerFactory logger,
+            [NotNull] UrlEncoder encoder,
+            [NotNull] ISystemClock clock)
+            : base(options, logger, encoder, clock)
+        {
+        }
+
+        protected override string BuildChallengeUrl(
+            [NotNull] AuthenticationProperties properties,
+            [NotNull] string redirectUri)
+        {
+            string stateValue = Options.StateDataFormat.Protect(properties);
+
+            string address = QueryHelpers.AddQueryString(Options.AuthorizationEndpoint, new Dictionary<string, string>
+            {
+                ["app_id"] = Options.ClientId,
+                ["redirect_uri"] = redirectUri,
+                ["scope"] = FormatScope(),
+                ["state"] = stateValue
+            });
+
+            return address;
+        }
+
+        protected override async Task<HandleRequestResult> HandleRemoteAuthenticateAsync()
+        {
+            var query = Request.Query.ToDictionary(c => c.Key, c => c.Value, StringComparer.OrdinalIgnoreCase);
+
+            if (query.TryGetValue("auth_code", out var authCode))
+            {
+                query["code"] = authCode;
+                Request.QueryString = QueryString.Create(query);
+            }
+
+            return await base.HandleRemoteAuthenticateAsync();
+        }
+
+        protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] OAuthCodeExchangeContext context)
+        {
+            var sortedParams = new SortedDictionary<string, string>()
+            {
+                ["app_id"] = Options.ClientId,
+                ["charset"] = "UTF-8",
+                ["code"] = context.Code,
+                ["format"] = "JSON",
+                ["grant_type"] = "authorization_code",
+                ["method"] = "alipay.system.oauth.token",
+                ["sign_type"] = "RSA2",
+                ["timestamp"] = Clock.UtcNow.ToString("yyyy-MM-dd HH:mm:ss", new CultureInfo("zh-CN")),
```

Comment:
```suggestion
                ["timestamp"] = Clock.UtcNow.ToString("yyyy-MM-dd HH:mm:ss", CultureInfo.InvariantCulture),
```

Reply:
`try.dot.net` renders this code `Console.WriteLine(DateTimeOffset.UtcNow.ToString("yyyy-MM-dd HH:mm:ss", new System.Globalization.CultureInfo("zh-CN")));` as `2020-02-15 17:14:00`, so there's no need to allocate a new `CultureInfo` that's specifically Chinese.

Reply:
Thanks.